### PR TITLE
[build.nims] New Builder for Arturo

### LINKED ISSUE #587: [build.nims] General cleanup needed
[build.nims] General cleanup needed

https://github.com/arturo-lang/arturo/blob/ce734e7a32c73173910cdf8f533f56be21febe3b/build.nims#L12

```text
# - Patrick (skydive241@gmx.de)

# TODO(build.nims) General cleanup needed
#  labels: installer, enhancement, cleanup

#=======================================
# Libraries
#=======================================
ndex 8b3a47db3..0363a98ed 100644
++ b/config.nims
```

b26a1368b631bfed1c96d4e575251970b5f87aa1

### PR DESCRIPTION
**Pull request by RickBarretto**
# Description

This PR is a trying of simplify our build script to something more maintainable and that generates fewer bugs along the time.

## Plans & Philosophies
* Decouple code
* Use metaprogramming when necessary for better APIs
* Better error messages
* Better defaults for developers and users (to discuss what should be the default)
* Better Cross-compilation
* ~Implement tests~

- Fixes #587, #588, #658

## Type of change

- [x] Code cleanup
- [ ] Unit tests (added or updated unit-tests)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (implementation update, or general performance enhancements)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (documentation-related additions)

### CODE REVIEW

**Comment by RickBarretto** [comment]
@drkameleon, I would like to hear your opinion about that before proceeding with that!

**Comment by drkameleon** [comment]
> @drkameleon, I would like to hear your opinion about that before proceeding with that!

To be honest: I absolutely love the idea.

Basically, do whatever you think is appropriate. If you can clean up all this mess (it has already been re-written, but as with all code that is patch upon patch, upon patch, to be honest I don't particularly love it...), it'll be more than great! 😉 

**My only to pre-requisites would be:**

- Everything that already works should keep working
- The less external tools we depend on (hence my obsession of not using `nimble`), the better. (Right now, the only thing that sb has to have installed before running the build script is Nim, that's all)

🚀 🚀 🚀

File: builder.nims [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    
+    ## This is an example of how we should write the tasks when it's finished
+        
+    var
+        
+        targetModes = some @["full", "mini", "web"]
+        targetCPUs  = some @["amd64", "arm", "arm64", "i386", "x86"]
+        targetOSes  = some @["freebsd", "linux", "openbsd", "macosx", "netbsd", "windows"]
+        
+        os:       OptionArg[string] = newOptionArg("os",       hostOS,              check=targetOSes)
```
Comment (RickBarretto) [inline builder.nims:26]: Or

```nim
config = BuildOptions()
# Searches for "--os <some string>" that matches with `targetOS`, if not found, returns the default.
config.os = args.getOptionParam(cmd="os", default=hostOS, in=targetOSes)
# Searches for "--install" or "-i" and returns true, if not found, returns the default.
config.install = args.hasFlag(cmd="install", default=off, short="i")
```

Comment (RickBarretto) [reply builder.nims:26]: For this, I would change the BuildOptions for something like

```nim
type BuildOptions = tuple
  os: string
  install: bool
```

**Comment by RickBarretto** [comment]
> The less external tools we depend on (hence my obsession of not using nimble), the better. (Right now, the only thing that sb has to have installed before running the build script is Nim, that's all)
🚀 🚀 🚀

I won't use `nimble`, I just see an use case if we put the [`extra`](https://github.com/arturo-lang/arturo/tree/master/src/extras)'s libraries into separated repositories (forked or remotes)...

**Comment by RickBarretto** [comment]
The original `task` did not work with optional arguments/flags, so I made `cmd` instead.
And IK, `hasFlag` and `getOptionValue` are both _O(n)_ algorithms per flag, but I don't see big problems since it's just a prototype and a simple CLI script.

If you think I should update this until the end of the PR, no worries, I can improve this algorithm soon! ;)

I'm trying to make it works first, to then improve it.

**Comment by RickBarretto** [comment]
How this is going on:
Code:
image
CLI:
image

I'll see if I found some bugs in this implementation, and also, add the @beef331's implementation for the automatic documentation, since mine has a very very slow algorithm.

**Comment by RickBarretto** [comment]
Updates! I've benchmarking different implementations of the `help()` function, and well... Here are the results:

## The first one

```nim
macro implementationToStr(id: typed): string =
    id.getImpl.toStrLit
  
proc getDocs(impl: string): seq[string] =
    result = @[]
    for line in impl.splitLines():
        let cleanline = line.strip()
        if cleanline.startsWith("##"):
            result.add cleanline[2 .. cleanline.high]
        # Don't iterate until the end of the implementation,
        # Just until the end of the impl's documentation
        elif result.len > 0:
            return
        
template help(ident: typed) =
  for line in ident.implementationToStr.getDocs():
    echo line
```

```
$ time ./builder.nims build --who bench --arch arm --help
...
real    0m5.824s
user    0m0.046s
sys     0m0.061s
```

## Using Iterator

```nim
macro implementationToStr(id: typed): string =
    id.getImpl.toStrLit
          
iterator getDocs(impl: string): string =
    var found = false
    for line in impl.splitLines():
        let cleanline = line.strip()
        if cleanline.startsWith("##"):
            yield cleanline[2 .. cleanline.high]
            found = true
        # Don't iterate until the end of the implementation,
        # Just until the end of the impl's documentation
        elif found:
            break
        
template help(ident: typed) =
  for line in ident.implementationToStr.getDocs():
    echo line
```

```
$ time ./builder.nims build --who bench --arch arm --help
...
real    0m5.716s
user    0m0.015s
sys     0m0.092s
```

## Jason's Implementation (@beef331)

```nim
macro help(t: proc): untyped =
    var comments: string
    let impl = t.getImpl
    
    if impl[^1].kind != nnkCommentStmt:
        for x in t.getImpl[^1]:
            if x.kind != nnkCommentStmt:
                break
            comments.add x.strVal & "\n"
    else:
        comments = impl[^1].strVal

    newCall "echo", newLit(comments)
```

```
$ time ./builder.nims build --who bench --arch arm --help
...
real    0m5.723s
user    0m0.015s
sys     0m0.092s
```

**Comment by RickBarretto** [comment]
## All available flags

I want to discuss which flags should be updated, what should be improved and why we need some of them.

```nim

## From build.nims

# Arch:

--cpu:arm                             # tags: arm32
--cpu:arm64                           # tags: arm64
--cpu:i386                            # tags: x86

--define:bit32                        # tags: arm32
--define:bit32                        # tags: x86

--gcc.path:/usr/bin                   # tags: arm64
--gcc.exe:aarch64-linux-gnu-gcc       # tags: arm64
--gcc.linkerexe:aarch64-linux-gnu-gcc # tags: arm64

--passC:'-m32'                        # tags: x86, gcc
--passL:'-m32'                        # tags: x86, gcc

# Developer

--embedsrc:on                         # tags: dev
--define:DEV                          # tags: dev
--listCmd                             # tags: dev

--define:DEBUG                        # tags: debug
--debugger:on                         # tags: debug
--debuginfo                           # tags: debug
--linedir:on                          # tags: debug

--define:PROFILE                      # tags: memprofile
--profiler:off                        # tags: memprofile
--stackTrace:on                       # tags: memprofile
--define:memProfiler                  # tags: memprofile

--define:PROFILE                      # tags: profiler
--profiler:on                         # tags: profiler
--stackTrace:on                       # tags: profiler

--debugger:native                     # tags: profilenative

# Compilation

--define:OPTIMIZED                    # tags: optimized

--define:strip                        # tags: release
--passC:-flto                         # tags: release, unix
--passL:-flto                         # tags: release, unix

# Compilation: full

--define:ssl                          # tags: full

# Compilation: mini/web

--define:SAFE                         # tags: safe
--define:MINI                         # tags: mini
--define:WEB                          # tags: web

--verbosity: 3                        # tags: web

--define:NOASCIIDECODE                # tags: noasciidecode
--define:NOCLIPBOARD                  # tags: noclipboard
--define:NODIALOGS                    # tags: nodialogs
--define:NOERRORLINES                 # tags: noerrorlines
--define:NOGMP                        # tags: nogmp
--define:NOPARSERS                    # tags: noparsers
--define:NOSQLITE                     # tags: nosqlite
--define:NOWEBVIEW                    # tags: nowebview

# Compilation: defaults
--verbosity:1                         # tags: default
--hints:on                            # tags: default
--hint:ProcessingStmt:off             # tags: default
--hint:XCannotRaiseY:off              # tags: default
--warning:GcUnsafe:off                # tags: default
--warning:CastSizes:off               # tags: default
--warning:ProveInit:off               # tags: default
--warning:ProveField:off              # tags: default
--warning:Uninit:off                  # tags: default
--warning:BareExcept:off              # tags: default
--threads:off                         # tags: default
--skipUserCfg:on                      # tags: default
--colors:off                          # tags: default
--define:danger                       # tags: default
--panics:off                          # tags: default
--mm:orc                              # tags: default
--define:useMalloc                    # tags: default            
--checks:off                          # tags: default
--cincludes:extras                    # tags: default
--opt:speed                           # tags: default
--nimcache:.cache                     # tags: default
--passL:-pthread                      # tags: default, windows-host
--path:src                            # tags: default
                                      # tags: default, windows
--passL:"-static-libstdc++ -static-libgcc -Wl,-Bstatic -lstdc++ -Wl,-Bdynamic" -- gcc.linkerexe='g++'
--passL:-lm                           # tags: default, unix

# Docs

--define:DOCGEN                       # tags: docs
--project                             # tags: docs
--index:on                            # tags: docs
--outdir:dev-docs                     # tags: docs

# Package

--forceBuild:on                       # tags: package
--opt:size                            # tags: package
--define:NOERRORLINES                 # tags: package
--define:PORTABLE                     # tags: package

## From config.nims

--cincludes:extras                    # tags: default
--path:src                            # tags: default
--hints:off                           # tags: default

                                      # tags: default, windows-host
--gcc.path:staticExec("pkg-config --libs-only-L gmp")
      .strip()
      .replace("-L","")
      .replace("/lib","/bin")
      .normalizedPath
      
--define:$mimallocStatic              # tags: default
--define:$mimallocIncludePath         # tags: default
                                      # tags: default, (gcc | clang | icc | icl)
--passC:"-ftls-model=initial-exec -fno-builtin-malloc"

--dynlibOverride:pcre64               # tags: default, windows
--define:noOpenSSLHacks               # tags: default, windows, ssl
--define:sslVersion:(                 # tags: default, windows, ssl
--dynlibOverride:ssl-                 # tags: default, windows, ssl
--dynlibOverride:crypto-              # tags: default, windows, ssl

--dynlibOverride:pcre                 # tags: default, macosx
--dynlibOverride:ssl                  # tags: default, unix, ssl
--dynlibOverride:crypto               # tags: default, unix, ssl
```

File: flags.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+--define:bit32                        # tags: arm32
+--define:bit32                        # tags: x86
+
+--gcc.path:/usr/bin                   # tags: arm64
+--gcc.exe:aarch64-linux-gnu-gcc       # tags: arm64
+--gcc.linkerexe:aarch64-linux-gnu-gcc # tags: arm64
+
+--passC:'-m32'                        # tags: x86, gcc
+--passL:'-m32'                        # tags: x86, gcc
```
Comment (RickBarretto) [inline flags.md:19]: `-m32` compiles in 32-bit mode.

Source: [GCC - x86 Options > `-m`][gcc]

[gcc]: https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#index-m32-2

File: builder.nims [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    ##     --who: string = user
+    ##          [bench, ci, dev, user]
+    ##     --compress -c
+    ##     --install -i
+    ##     --log -l
+    ##     --help
+
+    const
+        availableOSes  = @["freebsd", "linux", "openbsd", "macosx", "netbsd", "windows"]
+        availableCPUs  = @["amd64", "arm", "arm64", "i386", "x86"]
```
Comment (RickBarretto) [inline builder.nims:34]: - Idea: add `x86-64` and `x64` as `amd64` alias.

> In 1999–2003, [AMD](https://en.wikipedia.org/wiki/Advanced_Micro_Devices) extended this 32-bit architecture to 64 bits and referred to it as [x86-64](https://en.wikipedia.org/wiki/X86-64) in early documents and later as [AMD64](https://en.wikipedia.org/wiki/AMD64). Intel soon adopted AMD's architectural extensions under the name IA-32e, later using the name EM64T and finally using Intel 64. [Microsoft](https://en.wikipedia.org/wiki/Microsoft) and [Sun Microsystems](https://en.wikipedia.org/wiki/Sun_Microsystems)/[Oracle](https://en.wikipedia.org/wiki/Oracle_Corporation) also use term "x64", while many [Linux distributions](https://en.wikipedia.org/wiki/Linux_distribution), and the [BSDs](https://en.wikipedia.org/wiki/BSD) also use the "amd64" term. Microsoft Windows, for example, designates its 32-bit versions as "x86" and 64-bit versions as "x64", while installation files of 64-bit Windows versions are required to be placed into a directory called "AMD64".[[21]](https://en.wikipedia.org/wiki/X86#cite_note-32)
> - Wikipedia

source: https://en.wikipedia.org/wiki/X86

File: flags.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+--define:ssl                          # tags: full
+
+# Compilation: mini/web
+
+--define:SAFE                         # tags: safe
+--define:MINI                         # tags: mini
+--define:WEB                          # tags: web
+
+--verbosity: 3                        # tags: web
```
Comment (RickBarretto) [inline flags.md:61]: The idea of the new implementation is to `--verbosity` depend on the `--who` parameter.
By default, user should have `--verbosity:0`, while developers `--verbosity:3`.

* Should this be always 3 for web-builds?

File: flags.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+--define:PROFILE                      # tags: memprofile
+--profiler:off                        # tags: memprofile
+--stackTrace:on                       # tags: memprofile
+--define:memProfiler                  # tags: memprofile
+
+--define:PROFILE                      # tags: profiler
+--profiler:on                         # tags: profiler
+--stackTrace:on                       # tags: profiler
+
+--debugger:native                     # tags: profilenative
```
Comment (RickBarretto) [inline flags.md:41]: ## Current implementation

Questions:
* Why is `--profiler` is `off` for `memprofile`?
* What is the difference between the three approachs?

---

## New implementation

Questions:
* Should we add a `--profile` flag with the following options: `none`, `mem`, `default`, `native`?
* Should `--profile` be the default settings for `--who dev`? 
  * If yes, which of the options above should be the default one?

File: flags.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+# Compilation
+
+--define:OPTIMIZED                    # tags: optimized
+
+--define:strip                        # tags: release
+--passC:-flto                         # tags: release, unix
+--passL:-flto                         # tags: release, unix
```
Comment (RickBarretto) [inline flags.md:49]: This is the **link-time optimizer**.

source: [GCC - Optimize Options > `-flto`](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-flto)

File: flags.md (hunk 4 above)
Comment (drkameleon) [reply flags.md:61]: As a side-comment: web builds are generally *very* problematic. That's why it was set to `3`...

File: flags.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+--define:NOASCIIDECODE                # tags: noasciidecode
+--define:NOCLIPBOARD                  # tags: noclipboard
+--define:NODIALOGS                    # tags: nodialogs
+--define:NOERRORLINES                 # tags: noerrorlines
+--define:NOGMP                        # tags: nogmp
+--define:NOPARSERS                    # tags: noparsers
+--define:NOSQLITE                     # tags: nosqlite
+--define:NOWEBVIEW                    # tags: nowebview
+
```
Comment (RickBarretto) [inline flags.md:71]: All these `NO*`are for MINI and WEB builds.

File: flags.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+--hint:ProcessingStmt:off             # tags: default
+--hint:XCannotRaiseY:off              # tags: default
+--warning:GcUnsafe:off                # tags: default
+--warning:CastSizes:off               # tags: default
+--warning:ProveInit:off               # tags: default
+--warning:ProveField:off              # tags: default
+--warning:Uninit:off                  # tags: default
+--warning:BareExcept:off              # tags: default
+--threads:off                         # tags: default
+--skipUserCfg:on                      # tags: default
```
Comment (RickBarretto) [inline flags.md:84]: `SkipUserCfg` is essential to avoid collision with our compilation flags and don't mess things up.

Question:
* Should we also use `--skipParentCfg`?

source: https://nim-lang.org/docs/nims.html

File: flags.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+--colors:off                          # tags: default
+--define:danger                       # tags: default
+--panics:off                          # tags: default
+--mm:orc                              # tags: default
+--define:useMalloc                    # tags: default            
+--checks:off                          # tags: default
+--cincludes:extras                    # tags: default
+--opt:speed                           # tags: default
```
Comment (RickBarretto) [inline flags.md:92]: Question:
* Should speed be the default, or should this be used just for releases?
releases are for `user`, `bench` or `ci`.

If `--opt:speed` increases the compilation time, I think this would be a good idea to disable this for `--who dev`.

File: flags.md (hunk 4 above)
Comment (RickBarretto) [reply flags.md:61]: Alright!

File: flags.md (hunk 9 above)
Comment (RickBarretto) [reply flags.md:92]: @drkameleon, what do you think?

File: flags.md (hunk 8 above)
Comment (RickBarretto) [reply flags.md:84]: @drkameleon

**Comment by RickBarretto** [comment]
This will break the CI for while.
I renamed `config.nims` to `oldconfig.nims` and `builder.nims` to `config.nims`.

The reasons is at [rename builder.nims to config.nims][commit] commit.

[commit]: https://github.com/arturo-lang/arturo/pull/1287/commits/c87596ca0d829f85c4315b33ac7e1d637a99ea5e

File: config.nims [hunk 10]
```diff
@@ -1,49 +1,88 @@
...
+    ##     Provides a cross-compilation for the Arturo's binary.
+    ## 
+    ##     --arch: string = $hostCPU
+    ##          [amd64, arm, arm64, i386, x86]
+    ##     --build -b: string = full
+    ##          [full, mini, web]
+    ##     --os: string = $hostOS
+    ##          [freebsd, linux, openbsd, macosx, netbsd, windows]
+    ##     --who: string = user
+    ##          [bench, ci, dev, user]
```
Comment (RickBarretto) [inline config.nims:26]: `bench` should be the same as `ci`.

**Comment by RickBarretto** [comment]
Now the party will begin, with `builderConfig.nims` created, I need to call this from `builderUtils.nims`.

**Comment by RickBarretto** [comment]
Tested with `./builder build --who dev`, works pretty well!

Now, I need to do some changes, calibrate the logging, implement the other tasks, etc...

**Comment by RickBarretto** [comment]
Let's cross the fingers!

**Comment by RickBarretto** [comment]
`Permission denied`. Hmm...

**Comment by RickBarretto** [comment]
image
This is putting an `e` in the middle of the flags.

**Comment by RickBarretto** [comment]
On my local machine:

For Windows:
```bash
$ ./builder.nims build --who dev --log
build
>>> Compiling Arturo: @full

Building webview...

Flags:
        --path:src
        --cincludes:extras
        --nimcache:C:\msys64\home\barretto\arturo\.cache
        --skipUserCfg:on
        --skipParentCfg:on
        --colors:off
        --parallelBuild:0
        --define:strip
        --threads:off
        --mm:orc
        --define:useMalloc
        --define:mimallocStatic=C:\msys64\home\barretto\arturo\src\extras\mimalloc\src\static.c
        --define:mimallocIncludePath=C:\msys64\home\barretto\arturo\src\extras\mimalloc\include
        --verbosity:1
        --hints:on
        --define:danger
        --panics:off
        --checks:off
        --passL:-pthread
        --gcc.path:C:\msys64\mingw64\bin
        --hints:on
        --warnings:on
        --verbosity:3
        --hint:ProcessingStmt:off
        --hint:XCannotRaiseY:off
        --warning:GcUnsafe:off
        --warning:CastSizes:off
        --warning:ProveInit:off
        --warning:ProveField:off
        --warning:Uninit:off
        --warning:BareExcept:off
        --embedsrc:on
        --define:DEV
        --listCmd
        --define:DEBUG
        --debugger:on
        --debuginfo
        --linedir:on
        --gcc.linkerexe:g++
        --dynlibOverride:pcre64
        --passL:"-static-libstdc++ -static-libgcc -Wl,-Bstatic -lstdc++ -Wl,-Bdynamic"
        --define:noOpenSSLHacks
        --define:sslVersion:(
        --dynlibOverride:ssl-
        --dynlibOverride:crypto-
        --cpu:amd64
        --define:ssl


nim c --path:src --cincludes:extras --nimcache:C:\msys64\home\barretto\arturo\.cache --skipUserCfg:on --skipParentCfg:on --colors
:off --parallelBuild:0 --define:strip --threads:off --mm:orc --define:useMalloc --define:mimallocStatic=C:\msys64\home\barretto\a
rturo\src\extras\mimalloc\src\static.c --define:mimallocIncludePath=C:\msys64\home\barretto\arturo\src\extras\mimalloc\include --
verbosity:1 --hints:on --define:danger --panics:off --checks:off --passL:-pthread --gcc.path:C:\msys64\mingw64\bin  --hints:on --
warnings:on --verbosity:3 --hint:ProcessingStmt:off --hint:XCannotRaiseY:off --warning:GcUnsafe:off --warning:CastSizes:off --war
ning:ProveInit:off --warning:ProveField:off --warning:Uninit:off --warning:BareExcept:off --embedsrc:on --define:DEV --listCmd --
define:DEBUG --debugger:on --debuginfo --linedir:on --gcc.linkerexe:g++ --dynlibOverride:pcre64 --passL:"-static-libstdc++ -stati
c-libgcc -Wl,-Bstatic -lstdc++ -Wl,-Bdynamic" --define:noOpenSSLHacks --define:sslVersion:( --dynlibOverride:ssl- --dynlibOverrid
e:crypto- --cpu:amd64 --define:ssl --out:bin\arturo.exe src\arturo.nim
```

For Linux:
```
$ ./builder.nims build --who dev --os linux --log
build
>>> Compiling Arturo: @full

Building webview...

Flags:
        --path:src
        --cincludes:extras
        --nimcache:C:\msys64\home\barretto\arturo\.cache
        --skipUserCfg:on
        --skipParentCfg:on
        --colors:off
        --parallelBuild:0
        --define:strip
        --threads:off
        --mm:orc
        --define:useMalloc
        --define:mimallocStatic=C:\msys64\home\barretto\arturo\src\extras\mimalloc\src\static.c
        --define:mimallocIncludePath=C:\msys64\home\barretto\arturo\src\extras\mimalloc\include
        --verbosity:1
        --hints:on
        --define:danger
        --panics:off
        --checks:off
        --passL:-pthread
        --gcc.path:C:\msys64\mingw64\bin
        --hints:on
        --warnings:on
        --verbosity:3
        --hint:ProcessingStmt:off
        --hint:XCannotRaiseY:off
        --warning:GcUnsafe:off
        --warning:CastSizes:off
        --warning:ProveInit:off
        --warning:ProveField:off
        --warning:Uninit:off
        --warning:BareExcept:off
        --embedsrc:on
        --define:DEV
        --listCmd
        --define:DEBUG
        --debugger:on
        --debuginfo
        --linedir:on
        --passL:-lm
        --dynlibOverride:ssl
        --dynlibOverride:crypto
        --cpu:amd64
        --define:ssl


nim c --path:src --cincludes:extras --nimcache:C:\msys64\home\barretto\arturo\.cache --skipUserCfg:on --skipParentCfg:on --colors
:off --parallelBuild:0 --define:strip --threads:off --mm:orc --define:useMalloc --define:mimallocStatic=C:\msys64\home\barretto\a
rturo\src\extras\mimalloc\src\static.c --define:mimallocIncludePath=C:\msys64\home\barretto\arturo\src\extras\mimalloc\include --
verbosity:1 --hints:on --define:danger --panics:off --checks:off --passL:-pthread --gcc.path:C:\msys64\mingw64\bin  --hints:on --
warnings:on --verbosity:3 --hint:ProcessingStmt:off --hint:XCannotRaiseY:off --warning:GcUnsafe:off --warning:CastSizes:off --war
ning:ProveInit:off --warning:ProveField:off --warning:Uninit:off --warning:BareExcept:off --embedsrc:on --define:DEV --listCmd --
define:DEBUG --debugger:on --debuginfo --linedir:on --passL:-lm --dynlibOverride:ssl --dynlibOverride:crypto --cpu:amd64 --define
:ssl --out:bin\arturo.exe src\arturo.nim
```

I'm trying to know where that `e` came from.

**Comment by RickBarretto** [comment]
Interesting...

For web builds:

```
/home/runner/.choosenim/toolchains/nim-2.0.0/lib/pure/rationals.nim(103, 13) Error: illegal conversion from 'float' to 'int'
      ai = int(x)
              ^
??? compiler msg initiated here [MsgOrigin]
No stack traceback available
To create a stacktrace, rerun compilation with './koch temp js <file>', see https://nim-lang.github.io/Nim/intern.html#debugging-the-compiler for details
```

For Windows:

```
stack trace: (most recent call last)
D:\a\arturo\arturo\tools\builderUtils.nims(123, 24)
D:\a\arturo\arturo\builder.nims(47, 11) buildTask
D:\a\arturo\arturo\tools\builderUtils.nims(278, 16) buildArturo
D:\a\arturo\arturo\tools\builderConfig.nims(79, 13) buildConfig
D:\a\arturo\arturo\tools\builderConfig.nims(18, 12) strip
D:\a\arturo\arturo\tools\builderConfig.nims(18, 12) Error: index 1 not in 0 .. 0
Error: Process completed with exit code 1.
```

On MacOS:

```
# It compiles, but...
/Users/runner/work/_temp/ea9[8](https://github.com/arturo-lang/arturo/actions/runs/6757003945/job/18367067102?pr=1287#step:10:9)ac96-dff3-4273-888a-fcabeb7297ba.sh: line 2: arturo: command not found
Error: Process completed with exit code 127.
```

For Safe builds:

```
build
safe isn't into @["full", "mini", "web"] for --build.
Error: Process completed with exit code 1.
```

**Comment by RickBarretto** [comment]
Alright, I'll try it later... :/

**Comment by RickBarretto** [comment]
Sure! Trying a new approach, since errors are getting out of control...

For Linux, for instance, it's compiling, but when running tests:
```
>> Runtime | symbol not found: runSet
     error | perhaps you meant... truncate ?
           |                or... true ?
           |                or... request ?
```

For Web builds, I'm getting this from the standard library:
```
/home/runner/.choosenim/toolchains/nim-2.0.0/lib/pure/rationals.nim(103, 13) Error: illegal conversion from 'float' to 'int'
      ai = int(x)
              ^
```

For Mac, arturo is compiling, but not found...
```
/Users/runner/work/_temp/483a1087-768f-47dd-b51d-0610a4f6265e.sh: line 2: arturo: command not found
```

And for Windows a strange error that I had sometimes, when using `[1..^1]`, for instance, that unfortunately converts `-1` to a very high value, without reason... :/
```
D:\a\arturo\arturo\nim\nim-2.0.0\lib\system\indices.nim(85, 22) Error: illegal conversion from '-1' to '[0..9223372036854775807]'
```

## The new approach

Basically I'll take a lot of step backs. I'll maintain these new files here until the end of this PR for reference, but now, I'm going to re-factor instead of re-make the original one.
So, let's bring things as it was before, and change step-by-step until I like the final result taking the current implementation design as a goal.

**Comment by RickBarretto** [comment]
About the current error... Well, quantities related, so, I'm going to disable that test.
image

**Comment by drkameleon** [comment]
> About the current error... Well, quantities related, so, I'm going to disable that test. image

Spotted it just right now too:

https://github.com/arturo-lang/arturo/pull/1296/commits/b7824fb43d23f9dd4057de47d78c20ea08e6eb27

**Comment by RickBarretto** [comment]
@drkameleon, do you know what exactly would be pushing us this error?

```
>> Runtime | symbol not found: runSet
     error | perhaps you meant... truncate ?
           |                or... true ?
           |                or... request ?
```

**Comment by RickBarretto** [comment]
I know, I know... The code definitely does not look beautiful...
But I finally ran from string concatenation, and use the `--` template made by myself.
In the end this will make more sense, since my template `--` uses the same calling name of the original one that I plan to use in `config.nims`.

But first things first, I'm trying to slowly decouple the code. Now, the plan is to delete `OPTIONS` and use procedures, instead. Then, group common configurations, and then move everything to `config.nims`...

For now, the code still with a lot of repetition and using global variables...
But everything is working fine on CI, so no worries! 🙂

**Comment by drkameleon** [comment]
> @drkameleon, do you know what exactly would be pushing us this error?
> 
> ```
> >> Runtime | symbol not found: runSet
>      error | perhaps you meant... truncate ?
>            |                or... true ?
>            |                or... request ?
> ```
> 
> Update: Probably the binary was broken...

That one looks extremely weird to be honest - I've never come across something similar with the tester tool! Hmm...

Was it "solved" in the end?

**Comment by RickBarretto** [comment]
> > @drkameleon, do you know what exactly would be pushing us this error?
> > 
> > ```
> > >> Runtime | symbol not found: runSet
> >      error | perhaps you meant... truncate ?
> >            |                or... true ?
> >            |                or... request ?
> > ```
> > 
> > Update: Probably the binary was broken...
> 
> That one looks extremely weird to be honest - I've never come across something similar with the tester tool! Hmm...
> 
> Was it "solved" in the end?

Yes, this was solved. That was a strange... Behavior. I don't know exactly what happened.

**Comment by RickBarretto** [comment]
Now the fun begins!
I'll slowly change the CLI interface itself, and this will need some updated on workflow files...

**Comment by RickBarretto** [comment]
Next steps:
* organize the structure and helper functions
* split between behavior and UI focused procs
* eliminate global variables
* put it into `config.nims` (if you prefer)
  * or move `config.nims` to `.config/default.nims` or something else.

**Comment by RickBarretto** [comment]
@drkameleon, now for compiling in dev-mode, the usage will be to pass `--who dev` to the CLI.
Without the get environment's username...

**Comment by drkameleon** [comment]
> @drkameleon, now for compiling in dev-mode, the usage will be to pass `--who dev` to the CLI. Without the get environment's username...

Which makes perfect sense.

The check for whether the user is... some *drkameleon* shows how much I expected anybody else to play with the code! haha

Good job! 😉

**Comment by RickBarretto** [comment]
First time using Github's Codespace, well, great experience, bettern than I could imagine!

**Comment by RickBarretto** [comment]
@drkameleon 

I think this is good enough to be merged. 
Probably this needs some review. I'm finishing this now because I don't want to make this PR longer than it needs.

So, plans to use `config.nims` instead and reorganize even more our tasks would be good into a second PR maybe, if you think this is needed.

About the profiler, I did as it was before, you can check it in the `build.nims install`'s documentation for more information.

**Comment by drkameleon** [comment]
Jesus christ!
That's some insane amount of work! I've been looking into it for a while and I can say I'm amazed.

Let me have a more thorough look into it properly and I'll let you know.

Awesome job here - and at first sight, it looks far cleaner that we had! 🚀

**Review by drkameleon — CHANGES_REQUESTED** [review]
I finally got around to having a look into it as well! lol

And... I *do* like it (also: you might think that I'm easy to like... anything, but the truth is I'm totally not 😛 , so good job! 🚀 )

The code is much cleaner and nicely organized (I admit I'll need some time to get used to the new way it's organized, but as long as it works fine, I'm perfectly good with it). Plus, the more organized it is, the easier it'll be for us to resolve future issues. 😉 

One point (unless I completely missed it): in the previous version, there was an `--as:` option (basically, it allowed me to set the name (not the path, just the name) of the final binary to be built/installed to something other than Arturo. I've used this extensively so as to have different binaries which I can compare against (e.g. `arturo-1`, `arturo-2`, etc). Might sound like quite a simplistic approach, but it has worked for me quite well. Plus, I always keep the globally installed `arturo` clean, since I use it on a daily basis and I cannot afford e.g. to be working on a new not-so-functional feature and accidentally replace the binary.

Since I haven't seen that included, could you please add it back?

Once this is all set, I'll have another look into it and we're ready to go!

P.S. Of course, I might tweak a few things here and there in the future (especially things that you might have been unsure about - e.g. the mess with the one zillion "profile" modes, or the "package" mode - which has nothing to do with our package system and is not even properly working), but other than that: kudos! 😉 🚀

**Comment by RickBarretto** [comment]
> in the previous version, there was an --as: 

Yes, I can add this!

> Plus, I always keep the globally installed arturo clean, since I use it on a daily basis and I cannot afford e.g.

If I'm not wrong, I didn't change anything from the original one.

If you want, I had a plan to replace `install` by `build` command and make `--install|-i` a flag. What do you think?

**Comment by drkameleon** [comment]
> If you want, I had a plan to replace `install` by `build` command and make `--install|-i` a flag. What do you think?

You read my mind! 😄 

Yes, the way it's set up right now, it would make more sense! 😉

**Comment by RickBarretto** [comment]
Just a correction, the current way of preventing install globally is passing the `--local` flag.
But as you feel comfortable of me changing the `install` command by `build`, I'll do it, if you feel it's better with `--local` instead, I can take a step back.

**Comment by RickBarretto** [comment]
What do you think about this?
I mean, the only issue is the visual polution when I type `./build.nims build --build mini`...

Any suggestions?

**Comment by drkameleon** [comment]
> Just a correction, the current way of preventing install globally is passing the `--local` flag. But as you feel comfortable of me changing the `install` command by `build`, I'll do it, if you feel it's better with `--local` instead, I can take a step back.

I guess I had missed this part. The `--local` idea sound quite spot on to be honest. :)

**Comment by drkameleon** [comment]
> What do you think about this? I mean, the only issue is the visual pollution when I type `./build.nims build --build mini`...
> 
> Any suggestions? I would replace `--build` by `--mode` or `--version` if you want.

I do like your suggestion! I would go with `--mode` perhaps (`--version` would make someone believe it might have to do with a specific Arturo version or sth, which coulld be quite confusing..., no?)

**Review by drkameleon — CHANGES_REQUESTED** [review]
Once we get the two little details I commented above sorted above, I think we are ready to go.

Great job, @RickBarretto ! 😉 🚀

**Review by drkameleon — APPROVED** [review]
All set! Awesomeness! 😄

**Comment by drkameleon** [comment]
And...

Ready to merge! 🚀
